client/lib/purchases/index.ts

```diff
@@ -41,51 +41,165 @@ import type { ReactChild } from 'react';
 
 const debug = debugFactory( 'calypso:purchases' );
 
-export interface SiteWithPurchases {
+interface PurchaseWithStatus extends Purchase {
+	currentPurchaseStatus: string;
+}
+
+interface PurchaseStatus {
+	[ key: string ]: number;
+}
+
+interface SiteStatus {
+	[ key: string ]: number;
+}
+
+interface SiteWithPurchases {
 	id: number;
 	name: string;
 	slug: string;
 	isDomainOnly: boolean;
 	title: string;
-	purchases: Purchase[];
+	purchases: PurchaseWithStatus[];
+	isConnected: boolean;
 	domain: string;
+	currentStatus?: string;
 }
+
 export type TracksProps = Record< string, string | number | boolean >;
 
+// Site Sort order: (all purchases are grouped by site) - Sorted by if
+// the site or purchases of the site require some action.
+const siteStatus: SiteStatus = {
+	disconnected: 10,
+	pendingActivation: 20,
+	hasExpired: 30,
+	hasExpiringSoon: 40,
+	hasPaymentMethodExpired: 50,
+	noPaymentActionNeeded: 60,
+	hasCannotManage: 100,
+};
+
+// Sort order of the Purchases of each site (sorted by purchase status)
+const purchaseStatus: PurchaseStatus = {
+	expired: 10,
+	paymentMethodExpired: 20,
+	expiringSoon: 30,
+	noPaymentActionNeeded: 40,
+	cannotManage: 100,
+};
+
+function getSitePurchasesStatus( site: SiteWithPurchases ) {
+	if ( ! site.isConnected ) {
+		if ( site.slug === 'siteless.jetpack.com' ) {
+			return 'pendingActivation';
+		}
+		return 'disconnected';
+	}
+	const { purchases } = site;
+
+	if ( purchases.some( ( purchase ) => purchase.currentPurchaseStatus === 'cannotManage' ) ) {
+		return 'hasCannotManage';
+	}
+	if ( purchases.some( ( purchase ) => purchase.currentPurchaseStatus === 'expired' ) ) {
+		return 'hasExpired';
+	}
+	if (
+		purchases.some( ( purchase ) => purchase.currentPurchaseStatus === 'paymentMethodExpired' )
+	) {
+		return 'hasPaymentMethodExpired';
+	}
+	if ( purchases.some( ( purchase ) => purchase.currentPurchaseStatus === 'expiringSoon' ) ) {
+		return 'hasExpiringSoon';
+	}
+
+	return 'noPaymentActionNeeded';
```

Comment:
Effectively it'd do pretty much the same, but I'm wondering if this function and the one below should return `null` by default, instead of `'noPaymentActionNeeded'`.

Let's say we introduce a new value for `purchase.currentPurchaseStatus`. If we forget to update this function to handle that new status, the returned value `noPaymentActionNeeded` would be wrong, while `null` would give us a hint that we need to update the function.

Reply:
Thanks for the feedback @monsieur-z!  Hmm, if I'm understanding correctly, I don't think your suggestion will work for this scenario because the `noPaymentActionNeeded` has a _specific_ sort order position (that is not in the last position).
For example:
```
// The purchases sort order

const purchaseStatus: PurchaseStatus = {
	expired: 10,
	paymentMethodExpired: 20,
	expiringSoon: 30,
	noPaymentActionNeeded: 40,   // The 'noPaymentActionNeeded' position is Here!
	cannotManage: 100,
};
```

If we returned `null` in place of `noPaymentActionNeeded` then we would have to do something like this (which wouldn't work):

```
const purchaseStatus: PurchaseStatus = {
	expired: 10,
	paymentMethodExpired: 20,
	expiringSoon: 30,
	null: 40,   // ??? The  'noPaymentActionNeeded' position is Here! ???
	cannotManage: 100,
};
```

Let me know what you think. Or let me know if I am not understanding correctly. 😉